pkg/resources/ids.go

```diff
@@ -0,0 +1,176 @@
+	"fmt"
+	"strings"
+
+	"github.com/jszwec/csvutil"
+)
+
+const (
+	delimiter         = '|'
+	streamOndelimiter = '.'
+)
+
+func writeID(in []string) (string, error) {
+	var buf bytes.Buffer
+	csvWriter := csv.NewWriter(&buf)
+	csvWriter.Comma = delimiter
+	err := csvWriter.WriteAll([][]string{in})
+	if err != nil {
+		return "", err
+	}
+	strGrantID := strings.TrimSpace(buf.String())
+	return strGrantID, nil
+}
+
+func readID(id string, minFields, maxFields int) ([]string, error) {
+	reader := csv.NewReader(strings.NewReader(id))
+	reader.Comma = delimiter
+	lines, err := reader.ReadAll()
+	if err != nil {
+		return nil, fmt.Errorf("Not CSV compatible")
+	}
+
+	if len(lines) != 1 {
+		return nil, fmt.Errorf("expecting 1 line")
+	}
+	if len(lines[0]) < minFields || len(lines[0]) > maxFields {
+		if minFields == maxFields {
+			return nil, fmt.Errorf("%d fields allowed", minFields)
+		} else {
+			return nil, fmt.Errorf("between %d and %d fields allowed", minFields, maxFields)
+		}
+	}
+	return lines[0], nil
+}
+
+func readIDStruct(id string, data interface{}) error {
+	reader := csv.NewReader(strings.NewReader(id))
+	reader.Comma = delimiter
+
+	header, err := csvutil.Header(data, "csv")
+	if err != nil {
+		return err
+	}
+	fmt.Printf("[DEBUG] header %#v\n", header)
+	decoder, err := csvutil.NewDecoder(reader, header...)
+	if err != nil {
+		return err
+	}
+
+	return decoder.Decode(&data)
+}
+
+// grantID contains identifying elements that allow unique access privileges
+type grantID struct {
+	ResourceName string
+	SchemaName   string
+	ObjectName   string
+	Privilege    string
+	GrantOption  bool
+}
+
+// String() takes in a grantID object and returns a pipe-delimited string:
+// resourceName|schemaName|ObjectName|Privilege|GrantOption
+func (gi *grantID) String() (string, error) {
+	grantOption := fmt.Sprintf("%v", gi.GrantOption)
+	return writeID([]string{gi.ResourceName, gi.SchemaName, gi.ObjectName, gi.Privilege, grantOption})
+}
+
+// grantIDFromString() takes in a pipe-delimited string: resourceName|schemaName|ObjectName|Privilege
+// and returns a grantID object
+func grantIDFromString(stringID string) (*grantID, error) {
+	row, err := readID(stringID, 4, 5)
+	if err != nil {
+		return nil, err
+	}
+
+	grantOption := false
+	if len(row) == 5 && row[4] == "true" {
+		grantOption = true
+	}
+
+	grantResult := &grantID{
+		ResourceName: row[0],
+		SchemaName:   row[1],
+		ObjectName:   row[2],
+		Privilege:    row[3],
+		GrantOption:  grantOption,
+	}
+	return grantResult, nil
+}
+
+type schemaID struct {
+	Database string
+	Name     string
+}
+
+// String() takes in a schemaID object and returns a pipe-delimited string:
+// DatabaseName|schemaName
+func (si *schemaID) String() (string, error) {
+	return writeID([]string{si.Database, si.Name})
+}
+
+// schemaIDFromString() takes in a pipe-delimited string: DatabaseName|schemaName
+// and returns a schemaID object
+func schemaIDFromString(stringID string) (*schemaID, error) {
```

Comment:
🚫 **[golangci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>`schemaIDFromString` is unused (deadcode)